project/JavalibIRCleaner.scala

```diff
@@ -0,0 +1,435 @@
+package build
+
+import org.scalajs.ir._
+import org.scalajs.ir.ClassKind
+import org.scalajs.ir.Definitions._
+import org.scalajs.ir.Trees._
+import org.scalajs.ir.Types._
+
+import java.io._
+import java.nio.file.{Files, Path}
+import java.{util => ju}
+
+import scala.collection.mutable
+
+import sbt.{Logger, MessageOnlyException}
+
+/** Postprocessor for the IR of the javalanglib, to remove all references to
+ *  JS types in the Scala library, and ensure that there remains no other
+ *  reference to the Scala library.
+ *
+ *  This ensures that the IR of the javalanglib is truly independent of Scala.
+ *
+ *  The main task is to *completely* erase all references to JS types to
+ *  `j.l.Object`. This includes:
+ *
+ *  - Delete (or do not copy over) .sjsir files that define abstract and native
+ *    JS types.
+ *  - Erase references to JS types in method signatures and `TypeRef`s to
+ *    `java.lang.Object`.
+ *  - Eagerly dereference `LoadJSConstructor` and `LoadJSModule` by "inlining"
+ *    the JS load spec of the mentioned class ref.
+ *
+ *  Afterwards, we check that the IR does not contain any reference to classes
+ *  undef the `scala.*` package, with the unfortunate exception of
+ *  `scala.scalajs.runtime.UndefinedBehaviorError`.
+ */
+object JavalibIRCleaner {
+  def cleanIR(dependencyClasspath: Seq[File], javalibDir: File, outputDir: File,
+      logger: Logger): Set[File] = {
+
+    val errorManager = new ErrorManager(logger)
+
+    val javalibDirPath = javalibDir.toPath()
+    val outputDirPath = outputDir.toPath()
+
+    val libraryIRFiles = dependencyClasspath.flatMap(f => listIRFiles(f.toPath()))
+    val javalibIRFiles = listIRFiles(javalibDirPath)
+
+    val jsTypes = analyzeJSTypes(libraryIRFiles ++ javalibIRFiles)
```

Comment:
Don't know why I didn't point this out before: I find the name `analyze` strange here. Maybe just `getJSTypes` or `filter`/`find` `JSTypes`? Nothing is really analyzed :)